Sorry, please ignore - it still fails in a user project even with the snapshot, I tried something else.

So I've no ideas now what is going on - as this PR passes, locally, I've just retried, but fails here
with
```
Caused by: java.lang.NoClassDefFoundError: io/quarkus/jwt/test/User
```
even though it is added as a test class.

What can I do, it is Friday evening :-)

--- a/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesAllowedPostTest.java
+++ b/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesAllowedPostTest.java
@@ -1,0 +1,103 @@
+package io.quarkus.jwt.test;
+
+import static org.hamcrest.Matchers.equalTo;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.smallrye.jwt.build.Jwt;
+
+public class RolesAllowedPostTest {
+    private static Class<?>[] testClasses = {
+            RolesEndpoint.class,
+            User.class
+    };
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(testClasses)
+                    .addAsResource("publicKey.pem")
+                    .addAsResource("privateKey.pem")
+                    .addAsResource("applicationRolesAllowedPost.properties", "application.properties"));
+
+    @Test
+    public void postText() {
+        RestAssured.given().auth()
+                .oauth2(Jwt.upn("alice").groups("Tester").sign())
+                .when()
+                .body("principal:")
+                .post("/endp/postInjectedPrincipal")
+                .then()
+                .statusCode(200)
+                .body(equalTo("principal:alice"));
+    }
+
+    @Test
+    public void postTextInvalidToken() {
+        RestAssured.given().auth()
+                .oauth2("invalid")
+                .when()
+                .body("principal:")
+                .post("/endp/postInjectedPrincipal")
+                .then().statusCode(401);
+    }
+
+    @Test
+    public void postTextWrongRole() {
+        RestAssured.given().auth()
+                .oauth2(Jwt.upn("alice").groups("User").sign())
+                .when()
+                .body("principal:")
+                .post("/endp/postInjectedPrincipal")
+                .then().statusCode(403);
+    }
+
+    @Test
+    public void postJson() {
+        RestAssured.given()
+                .header("Content-Type", "application/json")
+                .auth().oauth2(Jwt.upn("alice").groups("Tester").sign())
+                .when()
+                .body("{\"name\":\"alice\"}")
+                .post("/endp/postInjectedPrincipalJson")
+                .then()
+                .statusCode(200)
+                .body(equalTo("name:alice,principal:alice"));
+    }
+
+    @Test
+    public void postJsonInvalidToken() {
+        RestAssured.given()
+                .header("Content-Type", "application/json")
+                .auth().oauth2("invalid")
+                .when()
+                .body("{\"name\":\"alice\"}")
+                .post("/endp/postInjectedPrincipalJson")
+                .then().statusCode(401);
+    }
+
+    @Test
+    public void postJsonNoToken() {
+        RestAssured.given()
+                .header("Content-Type", "application/json")
+                .when()
+                .body("{\"name\":\"alice\"}")
+                .post("/endp/postInjectedPrincipalJson")
+                .then().statusCode(401);
+    }
+
+    @Test
+    public void postJsonWrongRole() {
+        RestAssured.given()
+                .header("Content-Type", "application/json")
+                .auth()
+                .oauth2(Jwt.upn("alice").groups("User").sign())
+                .when()
+                .body("{\"name\":\"alice\"}")
+                .post("/endp/postInjectedPrincipalJson")
+                .then().statusCode(403);
+    }
+}

--- a/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesAllowedUnitTest.java
+++ b/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesAllowedUnitTest.java
@@ -18,6 +18,7 @@ public class RolesAllowedUnitTest {
             AuthenticatedEndpoint.class,
             PermitAllEndpoint.class,
             GreetingService.class,
+            User.class,
             TokenUtils.class
     };
     /**

--- a/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesEndpoint.java
+++ b/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/RolesEndpoint.java
@@ -8,7 +8,9 @@ import javax.annotation.security.PermitAll;
 import javax.annotation.security.RolesAllowed;
 import javax.enterprise.context.RequestScoped;
 import javax.inject.Inject;
+import javax.ws.rs.Consumes;
 import javax.ws.rs.GET;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.QueryParam;
 import javax.ws.rs.core.Context;
@@ -142,6 +144,21 @@ public class RolesEndpoint {
     public String getInjectedPrincipal(@Context SecurityContext sec) {
         boolean isJsonWebToken = this.jwtPrincipal instanceof JsonWebToken;
         return "isJsonWebToken:" + isJsonWebToken;
+    }
+
+    @POST
+    @Path("/postInjectedPrincipal")
+    @RolesAllowed("Tester")
+    public String postInjectedPrincipal(String body) {
+        return body + jwtPrincipal.getName();
+    }
+
+    @POST
+    @Path("/postInjectedPrincipalJson")
+    @RolesAllowed("Tester")
+    @Consumes("application/json")
+    public String postInjectedPrincipalJson(User user) {
+        return "name:" + user.getName() + ",principal:" + jwtPrincipal.getName();
     }
 
     @GET

--- a/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/User.java
+++ b/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/User.java
@@ -1,0 +1,15 @@
+package io.quarkus.jwt.test;
+
+public class User {
+
+    private String name;
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+}

--- a/extensions/smallrye-jwt/deployment/src/test/resources/applicationRolesAllowedPost.properties
+++ b/extensions/smallrye-jwt/deployment/src/test/resources/applicationRolesAllowedPost.properties
@@ -1,0 +1,8 @@
+mp.jwt.verify.publickey.location=/publicKey.pem
+mp.jwt.verify.issuer=https://server.example.com
+
+smallrye.jwt.sign.key.location=/privateKey.pem
+smallrye.jwt.new-token.issuer=https://server.example.com
+
+quarkus.log.category."io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator".min-level=TRACE
+quarkus.log.category."io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator".level=TRACE

--- a/integration-tests/smallrye-jwt-token-propagation/pom.xml
+++ b/integration-tests/smallrye-jwt-token-propagation/pom.xml
@@ -62,6 +62,23 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-jackson</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-jackson-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-oidc-token-propagation</artifactId>
         </dependency>
         <dependency>

--- a/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/AccessTokenPropagationService.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/AccessTokenPropagationService.java
@@ -1,6 +1,7 @@
 package io.quarkus.it.keycloak;
 
 import javax.ws.rs.GET;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 
 import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
@@ -14,4 +15,7 @@ public interface AccessTokenPropagationService {
 
     @GET
     String getUserName();
+
+    @POST
+    String echoUserName(String name);
 }

--- a/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/FrontendResource.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/FrontendResource.java
@@ -2,8 +2,10 @@ package io.quarkus.it.keycloak;
 
 import javax.annotation.security.RolesAllowed;
 import javax.inject.Inject;
+import javax.ws.rs.Consumes;
 import javax.ws.rs.GET;
 import javax.ws.rs.NotAuthorizedException;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 
 import org.eclipse.microprofile.config.inject.ConfigProperty;
@@ -41,6 +43,15 @@ public class FrontendResource {
     public String userNameAccessTokenPropagation() {
         checkIssuerAndAudience();
         return accessTokenPropagationService.getUserName();
+    }
+
+    @POST
+    @Path("access-token-propagation")
+    @RolesAllowed("user")
+    @Consumes("application/json")
+    public String userNameAccessTokenPropagationPost(User user) {
+        checkIssuerAndAudience();
+        return accessTokenPropagationService.echoUserName(user.getName());
     }
 
     private void checkIssuerAndAudience() {

--- a/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/ProtectedJwtResource.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/ProtectedJwtResource.java
@@ -1,7 +1,10 @@
 package io.quarkus.it.keycloak;
 
+import javax.annotation.security.RolesAllowed;
 import javax.inject.Inject;
+import javax.ws.rs.Consumes;
 import javax.ws.rs.GET;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.core.Context;
 import javax.ws.rs.core.SecurityContext;
@@ -26,12 +29,22 @@ public class ProtectedJwtResource {
 
     @GET
     @Path("test-security")
+    @RolesAllowed("viewer")
     public String testSecurity() {
         return securityContext.getUserPrincipal().getName();
     }
 
+    @POST
+    @Path("test-security")
+    @Consumes("application/json")
+    @RolesAllowed("viewer")
+    public String testSecurityJson(User user) {
+        return user.getName() + ":" + securityContext.getUserPrincipal().getName();
+    }
+
     @GET
     @Path("test-security-jwt")
+    @RolesAllowed("viewer")
     public String testSecurityJwt() {
         return accessToken.getName() + ":" + accessToken.getGroups().iterator().next()
                 + ":" + accessToken.getClaim("email");

--- a/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java
@@ -4,6 +4,7 @@ import javax.annotation.security.RolesAllowed;
 import javax.inject.Inject;
 import javax.ws.rs.GET;
 import javax.ws.rs.NotAuthorizedException;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 
 import org.eclipse.microprofile.config.inject.ConfigProperty;
@@ -27,6 +28,13 @@ public class ProtectedResource {
     public String principalName() {
         checkIssuerAndAudience();
         return jwt.getName();
+    }
+
+    @POST
+    @RolesAllowed("user")
+    public String echoPrincipalName(String name) {
+        checkIssuerAndAudience();
+        return name + ":" + jwt.getName();
     }
 
     private void checkIssuerAndAudience() {

--- a/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/User.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/main/java/io/quarkus/it/keycloak/User.java
@@ -1,0 +1,15 @@
+package io.quarkus.it.keycloak;
+
+public class User {
+
+    private String name;
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+}

--- a/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/KeycloakRealmResourceManager.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/KeycloakRealmResourceManager.java
@@ -35,6 +35,7 @@ public class KeycloakRealmResourceManager implements QuarkusTestResourceLifecycl
         realm.getClients().add(createClient("quarkus-app"));
         realm.getUsers().add(createUser("alice", "user"));
         realm.getUsers().add(createUser("bob", "user"));
+        realm.getUsers().add(createUser("john", "tester"));
 
         try {
             RestAssured

--- a/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/OidcTokenPropagationTest.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/OidcTokenPropagationTest.java
@@ -7,6 +7,7 @@ import org.junit.jupiter.api.Test;
 import io.quarkus.test.common.QuarkusTestResource;
 import io.quarkus.test.junit.QuarkusTest;
 import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
 
 @QuarkusTest
 @QuarkusTestResource(KeycloakRealmResourceManager.class)
@@ -28,5 +29,22 @@ public class OidcTokenPropagationTest {
                 .then()
                 .statusCode(200)
                 .body(equalTo("alice"));
+    }
+
+    @Test
+    public void testEchoUserNameWithAccessTokenPropagation() {
+        RestAssured.given().contentType(ContentType.JSON).auth().oauth2(KeycloakRealmResourceManager.getAccessToken("alice"))
+                .when().body("{\"name\":\"alice\"}").post("/frontend/access-token-propagation")
+                .then()
+                .statusCode(200)
+                .body(equalTo("alice:alice"));
+    }
+
+    @Test
+    public void testEchoUserNameWithAccessTokenPropagationForbidden() {
+        RestAssured.given().contentType(ContentType.JSON).auth().oauth2(KeycloakRealmResourceManager.getAccessToken("john"))
+                .when().body("{\"name\":\"alice\"}").post("/frontend/access-token-propagation")
+                .then()
+                .statusCode(403);
     }
 }

--- a/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java
+++ b/integration-tests/smallrye-jwt-token-propagation/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java
@@ -10,6 +10,7 @@ import io.quarkus.test.security.TestSecurity;
 import io.quarkus.test.security.jwt.Claim;
 import io.quarkus.test.security.jwt.JwtSecurity;
 import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
 
 @QuarkusTest
 @TestHTTPEndpoint(ProtectedJwtResource.class)
@@ -23,11 +24,31 @@ public class TestSecurityLazyAuthTest {
     }
 
     @Test
+    @TestSecurity(user = "user1", roles = "tester")
+    public void testWithDummyUserForbidden() {
+        RestAssured.when().get("test-security").then().statusCode(403);
+    }
+
+    @Test
+    @TestSecurity(user = "user1", roles = "viewer")
+    public void testPostWithDummyUser() {
+        RestAssured.given().contentType(ContentType.JSON).when().body("{\"name\":\"user1\"}").post("test-security").then()
+                .body(is("user1:user1"));
+    }
+
+    @Test
+    @TestSecurity(user = "user1", roles = "tester")
+    public void testPostWithDummyUserForbidden() {
+        RestAssured.given().contentType(ContentType.JSON).when().body("{\"name\":\"user1\"}").post("test-security").then()
+                .statusCode(403);
+    }
+
+    @Test
     @TestSecurity(user = "userJwt", roles = "viewer")
     @JwtSecurity(claims = {
             @Claim(key = "email", value = "user@gmail.com")
     })
-    public void testJwtWithDummyUser() {
+    public void testJwtGetWithDummyUser() {
         RestAssured.when().get("test-security-jwt").then()
                 .body(is("userJwt:viewer:user@gmail.com"));
     }

